File: client/beeper/src/logging.cpp

#include <Arduino.h>

#include "logging.h"

#define LOG_LEVEL_DISABLED	0
#define LOG_LEVEL_ERROR		1
#define LOG_LEVEL_WARN		2
#define LOG_LEVEL_INFO		3
#define LOG_LEVEL_DEBUG		4
#define LOG_LEVEL_TRACE		5

#define LOG_LEVEL LOG_LEVEL_DISABLED

namespace Logging {

	void init() {
#if LOG_LEVEL > LOG_LEVEL_DISABLED
		Serial.begin(115200);
#endif
	}

	void log(const char* log_level, const char* func_name, const char* msg) {
#if LOG_LEVEL > LOG_LEVEL_DISABLED
		Serial.printf("[%s] %s() - %s\n", log_level, func_name, msg);
#endif
	}


	void trace(const char* func_name, const char* msg) {
#if LOG_LEVEL >= LOG_LEVEL_TRACE
		Logging::log("TRACE", func_name, msg);
#endif
	}

	void debug(const char* func_name, const char* msg) {
#if LOG_LEVEL >= LOG_LEVEL_DEBUG
		Logging::log("DEBUG", func_name, msg);
#endif
	}

	void info(const char* func_name, const char* msg) {
#if LOG_LEVEL >= LOG_LEVEL_INFO
		Logging::log("INFO", func_name, msg);
#endif
	}

	void warn(const char* func_name, const char* msg) {
#if LOG_LEVEL >= LOG_LEVEL_WARN
		Logging::log("WARN", func_name, msg);
#endif
	}

	void error(const char* func_name, const char* msg) {
#if LOG_LEVEL >= LOG_LEVEL_ERROR
		Logging::log("ERROR", func_name, msg);
#endif
	}

	Trace::Trace(const char* func_name) {
#if LOG_LEVEL > LOG_LEVEL_DISABLED
		this->func_name = func_name;
		Logging::trace(this->func_name, ">");
#endif
	}
	
	Trace::~Trace() {
#if LOG_LEVEL > LOG_LEVEL_DISABLED
		Logging::trace(this->func_name, "<");
#endif
	}
};

File: server/logging.py

import inspect
import datetime

# Log levels
LOG_LEVEL_DISABLED = 0
LOG_LEVEL_ERROR    = 1
LOG_LEVEL_WARN     = 2
LOG_LEVEL_INFO     = 3
LOG_LEVEL_DEBUG    = 4
LOG_LEVEL_TRACE    = 5

# Set desired log level here
LOG_LEVEL = LOG_LEVEL_TRACE

def _caller_name(depth=2) -> str:
	frame = inspect.currentframe()
	try:
		return frame.f_back.f_back.f_code.co_name
	finally:
		del frame



class Logging:

	def _log(log_level: str, msg: str):
		if LOG_LEVEL > LOG_LEVEL_DISABLED:
			func_name = _caller_name()
			current_time = datetime.datetime.now().strftime("%d-%m-%Y %H:%M:%S")
			print(f"[{current_time}] [{log_level}] {func_name}() - {msg}", flush=True)


	def trace(msg: str):
		if LOG_LEVEL >= LOG_LEVEL_TRACE:
			Logging._log("TRACE", msg)


	def debug(msg: str):
		if LOG_LEVEL >= LOG_LEVEL_DEBUG:
			Logging._log("DEBUG", msg)


	def info(msg: str):
		if LOG_LEVEL >= LOG_LEVEL_INFO:
			Logging._log("INFO", msg)


	def warn(msg: str):
		if LOG_LEVEL >= LOG_LEVEL_WARN:
			Logging._log("WARN", msg)


	def error(msg: str):
		if LOG_LEVEL >= LOG_LEVEL_ERROR:
			Logging._log("ERROR", msg)



class Trace:

	def __enter__(self):
		if LOG_LEVEL > LOG_LEVEL_DISABLED:
			self.func_name = _caller_name(depth=3)
			print(f"[TRACE] {self.func_name}() - >")
		return self

	def __exit__(self, exc_type, exc_value, traceback):
		if LOG_LEVEL > LOG_LEVEL_DISABLED:
			print(f"[TRACE] {self.func_name}() - <")
		return False
